### ISSUE TITLE
Make Random.choice() method support multi-dimensional array

### ISSUE THREAD

**Issue by prashanth018** [chapel-lang/chapel#17136]
### Summary of Problem
Have been going through random module and observed that the choice() method only supports 1-D array. I will update the feature specific details soon, but on a bigger picture the changes to _choiceUniform() look straight forward while changes to _choiceProbabilities() will need some tweaking. 

### Steps to Reproduce

**Source Code:**

```chapel
use Random;
const vec = 1..7 by 2;
const vec1 = 2..8 by 3;
var D = {vec,vec1};
var A:[D] int;
A = 3;

var r = new owned PCGRandomStream(int);
writeln(r.choice(A));
```

**Compile command:**
`chpl foo.chpl`

**Output:**
Throws an error: says choice supports only 1-D array.

**Comment by prashanth018** [comment]
@lydia-duncan could you comment on this issue? I'll start working if its good to go. Thanks :)

**Comment by arjundashrath** [comment]
@prashanth018 @lydia-duncan can I work on it?

**Comment by prashanth018** [comment]
@arjundashrath I'm working on this already :)

**Comment by lydia-duncan** [comment]
@bradcray and @mppf - I think of you two as the experts on the Random module.  Does it seem reasonable to propose that the `choice` method should support multi-dimensional arrays?

**Comment by bradcray** [comment]
I've never really used `choice()`, so don't have any firsthand experience or motivators here, but I don't really see any reason not to support them (or even, for that matter, sparse or associative, though the implementation would be trickier and more expensive to maintain, so it may not be worth pursuing it until someone wants it).

**Comment by mppf** [comment]
maybe @ben-albrecht  has an opinion here?

**Comment by ben-albrecht** [comment]
> I don't really see any reason not to support them (or even, for that matter, sparse or associative, though the implementation would be trickier and more expensive to maintain, so it may not be worth pursuing it until someone wants it)

Agreed. I'm not opposed to adding support for N-D arrays. There just hasn't been user requests for it yet.

One design question for N-D arrays to consider: Should it sample in 1 dimension or all dimensions? 

By sampling along 1 dimension, I mean sample along 1 axis and return an `(N-1)`-dimensional array, e.g. a 2D array samples along rows and returns a 1D array. This is the behavior of [`numpy.random.Generator`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.choice.html#numpy.random.Generator.choice), which served as a reference for the 1D interface, but we can certainly diverge from that if we decide to sample all dimensions.

**Comment by lydia-duncan** [comment]
> One design question for N-D arrays to consider: Should it sample in 1 dimension or all dimensions?

This definitely seems like a good question to resolve before starting on the implementation :)

**Comment by prashanth018** [comment]
> > I don't really see any reason not to support them (or even, for that matter, sparse or associative, though the implementation would be trickier and more expensive to maintain, so it may not be worth pursuing it until someone wants it)
> 
> Agreed. I'm not opposed to adding support for N-D arrays. There just hasn't been user requests for it yet.
> 
> One design question for N-D arrays to consider: Should it sample in 1 dimension or all dimensions?
> 
> By sampling along 1 dimension, I mean sample along 1 axis and return an `(N-1)`-dimensional array, e.g. a 2D array samples along rows and returns a 1D array. This is the behavior of [`numpy.random.Generator.choice`](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.choice.html#numpy.random.Generator.choice), which served as a reference for the 1D interface, but we can certainly diverge from that if we decide to sample all dimensions.

Hi @ben-albrecht, The implementation I have in my mind is to sample it on N-dimensions. Lets say we have the below code.
```chapel
use Random;
const vec = 1..7 by 2; //size = 4
const vec1 = 2..8 by 3; //size = 3
const vec2 = 1..3; //size = 3
var D = {vec,vec1,vec2};
var C:[D] int;

forall (i,j,k) in D do
  C(i,j,k) = i*100 + j*10 + k;

var r = new owned PCGRandomStream(int);
writeln(r.choice(C, 4, false));
// prints 4 randomly chosen element across 3 dimensions
``` 
So this can be efficiently achieved by sampling a random number between 0 to C.size-1 and then using the [orderToIndex method](https://github.com/chapel-lang/chapel/blob/594f5de9e7f478b3e146ee776f2418b074eb6e2d/modules/internal/ChapelArray.chpl#L1898) of Chapel Array which returns the index given offset.

**Comment by ben-albrecht** [comment]
Thinking more on this, I think both approaches could be useful to include in the Random module. I would be OK with extending `choice` to support multi-dimensional arrays in the way you suggest @prashanth018 (assuming no others involved here object). I think the `N-1` dimension sampling could be implemented in a different function in the future.

**Comment by prashanth018** [comment]
I raised a [PR](https://github.com/chapel-lang/chapel/pull/17168) for this issue. It would be good if one of you shared your insights. Changes to choiceTestMultiDimensionalSupport.chpl file is not complete yet. I'm currently writing a utility file for tests. I'll be pushing these changes once the changes to random.chpl are confirmed. Thanks :)

**Comment by prashanth018** [comment]
I'll start working on the `N-1` dimension sampling that @ben-albrecht has suggested.

**Comment by prashanth018** [comment]
@lydia-duncan I'm planning to handle `N-1` sampling suggested by ben in a separate PR, if that is fine? Also, could you please review the changes? Thanks :)

**Comment by lydia-duncan** [comment]
I was expecting someone more familiar with the code to review the change

**Comment by lydia-duncan** [comment]
@ben-albrecht - would you mind looking at it?

**Comment by ben-albrecht** [comment]
I am a good person to look at it, but I won't get a chance to look until later this week (hopefully).

**Comment by prashanth018** [comment]
Thanks @lydia-duncan @ben-albrecht

**Comment by prashanth018** [comment]
Hey @ben-albrecht, would you mind taking a look at the changes if you find time? Thanks :)

**Comment by lydia-duncan** [comment]
@prashanth018 - note that @jeremiah-corrado resolved this issue with his #25512.  Thanks for reporting, and thanks Jeremiah for the fix!
